concordium-consensus/src/Concordium/Skov/Update.hs

```diff
@@ -165,34 +162,40 @@ processFinalization newFinBlock finRec@FinalizationRecord{..} = do
         -- Instead of marking blocks dead immediately we accumulate them
         -- and a return a list. The reason for doing this is that we never
         -- have to look up a parent block that is already marked dead.
-        let
-            pruneTrunk :: [BlockPointerType m] -- ^List of blocks to remove.
+        let pruneTrunk :: [BlockPointerType m] -- ^List of blocks to remove.
                        -> BlockPointerType m -- ^The block that was finalized.
                        -> Branches m
-                       -> m [BlockPointerType m]
+                       -> m ([BlockPointerType m], [BlockPointerType m])
                        -- ^ The return value is a list of blocks to mark dead, ordered
                        -- by increasing height.
-            pruneTrunk toRemove _ Seq.Empty = return toRemove
+            pruneTrunk toRemove _ Seq.Empty = return ([], toRemove)
             pruneTrunk toRemove keeper (brs Seq.:|> l) = do
                 parent <- bpParent keeper
                 let toRemove1 = filter (/= keeper) l ++ toRemove
-                toRemove2 <- pruneTrunk toRemove1 parent brs
-                -- mark blocks as finalized now, so that blocks are marked finalized by increasing height
-                markFinalized (getHash keeper) finRec
-                logEvent Skov LLDebug $ "Block " ++ show keeper ++ " marked finalized"
-                -- Finalize the transactions of the surviving block.
-                -- (This is handled in order of finalization.)
-                finalizeTransactions (getHash keeper) (blockSlot keeper) (blockTransactions keeper)
-                ati <- bpTransactionAffectSummaries keeper
-                bcTime <- getSlotTimestamp (blockSlot keeper)
-                let ctx = BlockContext{
-                        bcHash = getHash keeper,
-                        bcHeight = bpHeight keeper,
-                        ..}
-                flushBlockSummaries ctx ati =<< getSpecialOutcomes =<< blockState keeper
-                return toRemove2
-
-        toRemoveFromTrunk <- pruneTrunk [] newFinBlock (Seq.take pruneHeight oldBranches)
+                (toFinalize, toRemove2) <- pruneTrunk toRemove1 parent brs
+                return (keeper : toFinalize, toRemove2)
+        (toFinalize, toRemoveFromTrunk) <- pruneTrunk [] newFinBlock (Seq.take pruneHeight oldBranches)
```

Comment:
If I'm understanding this correctly then `toFinalize` is a list of blocks "by decreasing height".

`finalizeTransactions` assumes that all transactions prior to the block it is processing are finalized. Thus this will go very badly.

Either reverse the `toFinalize` list here, or construct it in reverse order via an accumulator.

If you do the latter, you can also construct `toRemoveFromTrunk` in reverse order, avoiding reversal below, and making `pruneTrunk` tail recursive.

Reply:
And the comment of pruneTrunk should probably be updated to reflect the new reality.

Reply:
I believe that `toFinalize` was already constructed in __increasing__ order but, indeed, it's easier to see that if `pruneTrunk` uses an accumulator to construct it: `parent` is always prepended before `toFinalize`.

I also reversed the order in which `toRemove` is constructed, it's in decreasing order now.